pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1975,6 +1982,9 @@ public void seek(Function<String, Object> function) throws PulsarClientException
                 ackSet.recycle();
 
                 seek = Commands.newSeek(consumerId, requestId, msgId.getLedgerId(), msgId.getEntryId(), ackSetArr);
+            } else if(messageId instanceof ChunkMessageIdImpl){
+                ChunkMessageIdImpl msgId = (ChunkMessageIdImpl) messageId;
+                seek = Commands.newSeek(consumerId, requestId, msgId.getFirstChunkMessageId().getLedgerId(), msgId.getFirstChunkMessageId().getEntryId(), new long[0]);
```

Comment:
this change is submitted and new message-type `ChunkMessageIdImpl` is introduced based on consideration of the presence of `firstChunkMessageId` and that will not be present for most of the time. so, this will not work as expected.

Reply:
Because chunk messages are sent with guaranteed order for each chunk, the first chunk message id here will not be null.

Reply:
> Because chunk messages are sent with guaranteed order for each chunk

Yes, ordering is guaranteed. but it's not guaranteed that you will get publish ACK of first chunk before you publish last chunk. It seems I am explaining the same thing again and again. Am I really missing anything here?

Reply:
>  but it's not guaranteed that you will get publish ACK of first chunk before you publish last chunk.

As I said earlier. We don't need that guarantee. I think this is the confusing point of the discussion. The firstChunkMessageId that the consumer gets at the time of seek is not given by the producer, but is calculated at the consumer side.

Reply:
the fundamental assumption at the consumer side is : `msgId.getFirstChunkMessageId().getLedgerId()` will be present. and this assumption is incorrect. `msg -> getFirstChunkMessageId()` can be null because what I explained earlier: `you can't give guarantee to set firstChunkMessageId in any chunk of the message.`

Reply:
As I explained earlier([here](https://github.com/apache/pulsar/pull/12403#discussion_r756558399) and [here](https://github.com/apache/pulsar/pull/12403#discussion_r747272455)), the Consumer does not depend on the first chunk message-id set by the producer. [Here](https://github.com/apache/pulsar/pull/12403/files#r759004547) is the comment to illustrate where the consumer sets the chunk message-id. Is there something wrong with my explanation?

> `you can't give guarantee to set firstChunkMessageId in any chunk of the message.`

Why do we need that guarantee? I think this is the key point of our discussion. The consumer does not rely on this guarantee.

Reply:
if consumer doesn't depend on first chunk id then why do we have seek method in consumerImpl on this line which uses firstChunkMessageId?
```
seek = Commands.newSeek(consumerId, requestId, msgId.getFirstChunkMessageId().getLedgerId(), msgId.getFirstChunkMessageId().getEntryId(), new long[0]);
```

not sure what are we trying to achieve in this PR.

Reply:
The consumer doesn't depend on the first chunk id **set by the producer**, but does depend on the first chunk id **calculated by the consumer itself**. If you go through the code, you can find that this first chunk id here is either calculated by the producer or by the consumer. If this chunk msg id is from the `consumer.receive`, then the first chunk id here is calculated by the consumer but not the producer.